### PR TITLE
GLES2 batching - Add UV precision adjustment for tilemaps

### LINKED ISSUE #39096: 1px gaps between tiles on Android
**Note:** This issue is created after a discussion in #38004, where the initial assumption was that this problem was due to the new batching feature in 3.2.2. Links to the posts are included at the end of this issue as reference.

**Godot version:**
3.2.2 beta3 official

**OS/device including version:**
Galaxy Tab A (SM-T510), 1920x1200px resolution, Android 9,
GPU: Mali-G71 MP2
Backend: GLES2

**Issue description:**
While trying out v3.2.2 beta3 I noticed on one of my Android devices that in some instances there are visible 1px gaps between tiles in my tilemaps, see the following screenshot. When disabling batching the gaps are gone.
batching_bug_minimal

* The gaps disappear when switching of batching.
* The gaps only appears for tiles that are created with an offset inside a texture, i.e. a 32x32px tile created from a 32x32px image will not have this problem, but a 32x32px tile created at row 256 in a 32x320px image will have the problem.
* The gaps are only visible on my android tablet, not my PC, even if I set the window size to the same resolution as my tablet.
* Filter is disabled for my tile texture, pixel snap is enabled in project settings.
* Changing the project window size has an effect on the artifacts, e.g. 320x224 shows the problem while 320x400 does not.

**Steps to reproduce:**
* Create a new project.
* Set project resolution to 320x224, stretch mode to 2d, stretch aspect to keep, enable pixel snap
* Add a new tilemap.
* Add a new tileset to the tilemap.
* Using the texture from the attached minimal example project, create a new 32x32 tile at position x:256 y:0
* Draw a few tiles
* Run the project on Galaxy Tab A

**Minimal reproduction project:**
[batching_bug_minimal.zip](https://github.com/godotengine/godot/files/4690777/batching_bug_minimal.zip)

### Previous posts on this:
https://github.com/godotengine/godot/issues/38004#issuecomment-634097542
https://github.com/godotengine/godot/issues/38004#issuecomment-634545040
https://github.com/godotengine/godot/issues/38004#issuecomment-634581064
https://github.com/godotengine/godot/issues/38004#issuecomment-634590975
https://github.com/godotengine/godot/issues/38004#issuecomment-634743475
https://github.com/godotengine/godot/issues/38004#issuecomment-634750388
https://github.com/godotengine/godot/issues/38004#issuecomment-634781122

### PR DESCRIPTION
**Pull request by lawnjelly**
Scaling tilemaps can cause border artifacts around the edges of tiles. This has been traced to precision issues in the GPU. This PR adds an adjustment to allow a minor contraction of the UVs of rects in order to compensate for the incorrect classification of texels across the UV border.

Fixes #39096.

### Notes
* See the issue for a detailed description of the problem and approaches to solving it
* This problem seems less likely to occur using the old uniform draw method. However it still exists - border artifacts occur when pixel_snap is off even with uniform draw. Also there are reports of this occurring pre-batching.
* This problem is very clear with either the nvidia_workaround, or the batching draw methods, which are standard drawing methods passing regular vertices.
* Other engines may deal with this problem by adding a single pixel border around tiles on tilemaps. While it may be worth adding this capability, we can't rely on this for backward compatibility reasons.
* The contraction may lead to a corresponding decrease in the size of the edge texels at high magnifications. This is undesirable, but may be a cost of dealing with this problem.
* There is a setting to enable / disable the feature. I did consider only having it on with pixel_snap, however the problem occurs with pixel_snap off as well so it could benefit from independent control.
* The offset is applied only when using unfiltered textures (i.e. GL_NEAREST)
* I have used the new 'batching' common project setting, in accordance with #39068 (moving common batching settings into a batching folder, rather than gles2) - these adjustments will also be needed in GLES3.

### Defaults
After further thought I decided to add a separate on / off setting and amount setting, and the default is off. There could be an argument for having it on as default - I'll leave this to the reviewers.

### Padded tilemaps
I am also investigating adding the ability to Godot to automatically create and use padded tilemaps from unpadded source tilemaps. Although this may add one more (possibly automatic) step to the tilemap creation process, it may be the only way to solve the texel precision problem in an efficient manner with no squashing of edge pixels. However it will not be backward compatible, and would have to be applied to all atlased artwork (including animated sprites), so the technique in this PR would probably still be needed.

### CODE REVIEW

**Comment by mattislorentzon** [comment]
I just wanted to confirm that this patch has the same effect on my device as in the intermediate patch evaluated here:
https://github.com/godotengine/godot/issues/39096#issuecomment-636899670
(the one using `max_join_item_commands = 45`)

Great work! :partying_face:

**Comment by lawnjelly** [comment]
> I just wanted to confirm that this patch has the same effect on my device as in the intermediate patch evaluated here:
> [#39096 (comment)](https://github.com/godotengine/godot/issues/39096#issuecomment-636899670)
> (the one using `max_join_item_commands = 45`)

Yes, the default value of 45 is straight from your experiments. :+1: However it is still possible it may need changing to a slightly larger figure after more testing, because the value may be dependent on the relative tile and tilemap dimensions.

**Review by clayjohn — COMMENTED** [review]
Looks good. 

With this in place, can we remove the hack in canvas.glsl?

**Comment by lawnjelly** [comment]
> Looks good.
> 
> With this in place, can we remove the hack in canvas.glsl?

I did a lot of thinking about that. The original version of this PR actually had an offset to correct for the hack in canvas.glsl, but then I realised that if the hack was actually doing a useful job (giving an absolute offset and helping to classify texels _within_) it might still be needed. Also the shader hack works on all primitives, and this UV contraction is only working on rects.

At this stage I'm not absolutely sure we can remove the shader hack - without having been able to replicate the other effect myself, I can't confirm or deny whether it is also solved by UV contraction, so I've left it as is for now.

I'm actually kind of amazed the other effect happens at all, given that precision issues should be identical across a horizontal / vertex line, IF there is no rotation. Even with a rotation I'd expect there just to be a single break in lines, not a jittering effect. I.e. precision errors should be systematic rather than 'random'.

File: doc/classes/ProjectSettings.xml [hunk 1]
```diff
@@ -970,6 +970,12 @@
 			Fix to improve physics jitter, specially on monitors where refresh rate is different than the physics FPS.
 			[b]Note:[/b] This property is only read when the project starts. To change the physics FPS at runtime, set [member Engine.physics_jitter_fix] instead.
 		</member>
+		<member name="rendering/batching/precision/uv_contract" type="bool" setter="" getter="" default="false">
+		On some platforms (especially mobile), precision issues in shaders can lead to reading 1 texel outside of bounds, particularly where rects are scaled. This can particularly lead to border artifacts around tiles in tilemaps. This adjustments corrects for this by making a small contraction to the UV coordinates used. Note that this can result in a slight squashing of border texels.
```
Comment (akien-mga) [inline doc/classes/ProjectSettings.xml:974]: Indentation issue (missing one tab).

Comment (akien-mga) [reply doc/classes/ProjectSettings.xml:974]: Maybe add a mention to `[member rendering/quality/2d/use_pixel_snap]` and mention this new setting in `use_pixel_snap`'s description, as I guess they're related and will often be needed together?

Comment (lawnjelly) [reply doc/classes/ProjectSettings.xml:974]: Good catch, yup I'll add something. :+1:

**Comment by robertarnborg** [comment]
Hope this can fix my issues I had. Otherwise I'll put a polygon 2d behind my tiles 😆

Good work Lawnjelly

**Pull request merged**